### ISSUE TITLE
It would be better if vim's virtual text can overlay the actual text.

### ISSUE THREAD

**Issue by Yggdroot** [vim/vim#15007]
neovim has `nvim_buf_set_extmark` to create virtual text, it supports the "overlay" virtual text positon.
```
                  • virt_text_pos : position of virtual text. Possible values:
                    • "eol": right after eol character (default).
                    • "overlay": display over the specified column, without
                      shifting the underlying text.
                    • "right_align": display right aligned in the window.
                    • "inline": display at the specified column, and shift the
                      buffer text to the right as needed.
```

vim's `prop_add` don't support the "overlay" virtual text position.

**Comment by chrisbra** [comment]
so what does overlay do?

**Comment by Yggdroot** [comment]
https://github.com/easymotion/vim-easymotion
https://github.com/hadronized/hop.nvim

vim-easymotion is a famous plugin used in vim, and hop.nvim is an alternative in neovim.

As the movivation of hop.vim says below:

## Motivation
Hop is a complete from-scratch rewrite of [EasyMotion](https://github.com/easymotion/vim-easymotion), a famous plugin to enhance the native motions of Vim. Even though [EasyMotion](https://github.com/easymotion/vim-easymotion) is usable in Neovim, it suffers from a few drawbacks making it not comfortable to use with Neovim version >0.5 – at least at the time of writing these lines:

[EasyMotion](https://github.com/easymotion/vim-easymotion) uses an old trick to annotate jump targets by saving the contents of the buffer, replacing it with the highlighted annotations and then restoring the initial buffer after jump. This trick is dangerous as it will change the contents of your buffer. A UI plugin should never do anything to existing buffers’ contents.
Because the contents of buffers will temporarily change, other parts of the editor and/or plugins relying on buffer change events will react and will go mad. An example is the internal LSP client implementation of Neovim >0.5 or its treesitter native implementation. For LSP, it means that the connected LSP server will receive a buffer with the jump target annotations… not ideal.
Hop is a modern take implementing this concept for the latest versions of Neovim.



**The drawback of EasyMotion in Vim is that virtual text cannot be overlaid on top of the actual text. However, Neovim does support this feature.**

https://github.com/lukas-reineke/indent-blankline.nvim also uses this feature.

**Comment by chrisbra** [comment]
so, can you show an example/screenshot what exactly this does?

**Comment by Yggdroot** [comment]
![2](https://github.com/vim/vim/assets/2705841/a126b669-5a24-478d-9d95-69264e580987)

**Comment by bfrg** [comment]
Have you tried using popup windows for that?

**Comment by Yggdroot** [comment]
> Have you tried using popup windows for that? I know it's a little bit weird. I would also welcome this feature.

No. I think using popups has limitations and the performance is somewhat low in this scenario.

**Comment by vimpostor** [comment]
> Have you tried using popup windows for that?

I have already created a plugin for this, it works reasonably well (showing the hints with `popup_create()`: https://github.com/vimpostor/vim-gallop

I have benchmarked this a little bit and indeed, under extreme circumstances vim is significantly slower than neovim due to the overhead of popups being actual windows (vim on the left, neovim on the right):

[popups.webm](https://github.com/user-attachments/assets/a7944bbb-c958-48eb-ae8b-dccb753f9688)

In this benchmark I show 4096 popups at once (these are the hints covering the entire screen). Granted, this is an extreme case, but it makes benchmarking more obvious. For the neovim implementation I use `extmarks` with `virt_text_pos` set to `overlay`. For vim I use `popup_create()`, because no other `overlay` implementation is available.
You can clearly see in the video how much slower Vim is compared to Neovim, the results speak for themselfes:
```
Vim: 3.3s
Neovim: 0.1s
```

Now coming back to the real world usecase of showing hints in a window, under normal operation the performance difference is not really noticable, but if I maximise the window on a 4k screen, one can already slightly feel the lag for vim, while the neovim implementation shows all hints instantly.

@chrisbra Would you accept a patch that adds `overlay` support to `virtual-text`, if I can prove that this solves the performance problem for those usecases, where one just needs to overlay some text?

**Comment by chrisbra** [comment]
> @chrisbra Would you accept a patch that adds overlay support to virtual-text, if I can prove that this solves the performance problem for those usecases, where one just needs to overlay some text?

Sure.

**Comment by qeatzy** [comment]
@bfrg @vimpostor @Yggdroot The performance of popup is for this (EasyMotion like hint mode) is in fact extremely well, even under most extreme cases (but with one caveat). Instead of use a lot of windows, it can always be done with less than one hundred popop windows (less than vim screen height). All you need is using popup mask feature. I've done this before and can conform it is instantaneous (less than 50ms).

The tips for performance.
1. do less work, do not filter what you don't need (those invisible text).
2. pre-create popup windows (hide & show instead of close-recreate, unless closed by others, eg popup_clear() etc)
3. use popup mask to reduce number of popup windows (you need one popup for screen line 3 even if has tens of matches). Which do require correct result of screenpos().
4. use vim9script or luajit to implement matching part.
5. do pre-filter before user typed any char. (eg, find all case of '[a-z][a-z]' etc when waiting key press)

In order of performance impact, from high to low
- for extreme case 2 > 3 > 4 > 5 > 1.
- for normal case 3 > 2 == 5 > 4  > 1.
